# STA 663 Project 1: 
- Huggingface: mastergopote44/Long-Term-Care-Aggregated-Data
- Name: Justin Kao
- School: Duke University

## Load Incidence & Termination dataset from SOA website

In [1]:
import requests
from io import BytesIO
from zipfile import ZipFile
import pandas as pd

# Directory to store the extracted files
desired_directory = '/Users/justinkao/Desktop/Durham NC/Duke University/Courses/S_2024_STA 663 Statistical Computing and Computation(Dr. Ouwen Huang)/Project 1/STA663_Project_1'  # Replace with your actual path

# URLs of the zip files
zip_files = {
    'incidence': 'https://www.soa.org/4a33aa/globalassets/assets/files/resources/experience-studies/2020/2000-2016-ltc-incidence.zip',
    'termination': 'https://www.soa.org/4a2e5d/globalassets/assets/files/resources/experience-studies/2020/2000-2016-ltc-termination.zip'
}

# DataFrames dictionary
dataframes = {}

# Download and unzip the files
for name, zip_url in zip_files.items():
    response = requests.get(zip_url)
    # Check if the request was successful
    if response.ok:
        with ZipFile(BytesIO(response.content)) as thezip:
            # Extract all files from the zip into a directory
            thezip.extractall(desired_directory)
            # Loop through each file in the zip
            for zipinfo in thezip.infolist():
                # Construct the full path to the extracted file
                file_path = f'{desired_directory}/{zipinfo.filename}'
                # Assuming the file is a tab-separated txt file
                df = pd.read_csv(file_path, sep='\t', low_memory=False)  # Adjust sep if necessary
                # Assign the DataFrame to the corresponding variable
                dataframes[name] = df
    else:
        print(f"Failed to retrieve {zip_url}")

In [2]:
incidence_df = dataframes['incidence']
incidence_df

Group_Indicator  Gender Issue_Age_Bucket Incurred_Age_Bucket  \
0                 Group  Female            55-59               55-59   
1            Individual  Female            60-64               70-74   
2                 Group  Female            50-54               60-64   
3            Individual  Female            55-59               70-74   
4            Individual    Male            60-64               65-69   
...                 ...     ...              ...                 ...   
2089318      Individual    Male            55-59               60-64   
2089319      Individual    Male            65-69               70-74   
2089320      Individual  Female            65-69               75-79   
2089321      Individual  Female            50-54               55-59   
2089322      Individual  Female             0-49               50-54   

        Issue_Year_Bucket  Policy_Year Marital_Status Premium_Class  \
0               2003-2005    1-3 years         Single      Standard   
1               1997-1999  10-12 years         Single     Preferred   
2               2000-2002  10-12 years        Unknown      Standard   
3               1994-1996  13-15 years        Married      Standard   
4               2003-2005    4-6 years        Married   Substandard   
...                   ...          ...            ...           ...   
2089318         2012-2014    1-3 years        Married   Substandard   
2089319         1997-1999    7-9 years        Married      Standard   
2089320         1994-1996    7-9 years         Single     Preferred   
2089321         1994-1996    7-9 years         Single      Standard   
2089322         1997-1999    1-3 years         Single     Preferred   

         Underwriting_Type Coverage_Type_Bucket  ... ALF_EP_Bucket  \
0                    Other        Comprehensive  ...             0   
1                    Other        Comprehensive  ...             0   
2                  Unknown        Comprehensive  ...             0   
3                    Other        Comprehensive  ...             0   
4        Full underwriting        Comprehensive  ...            90   
...                    ...                  ...  ...           ...   
2089318  Full underwriting        Comprehensive  ...             0   
2089319  Full underwriting        Comprehensive  ...           180   
2089320              Other        Comprehensive  ...             0   
2089321            Unknown        Comprehensive  ...             0   
2089322  Full underwriting        Comprehensive  ...            90   

        HHC_EP_Bucket         Region Active_Exposure Total_Exposure  \
0                   0        Unknown       37.583332      37.583332   
1                   0  02: Northeast       14.500000      14.500000   
2                   0        Unknown     2250.000000    2252.000000   
3                   0   01: Mid-West       19.000000      19.000000   
4                  20  02: Northeast        2.000000       2.000000   
...               ...            ...             ...            ...   
2089318             0   01: Mid-West        7.083333       7.083333   
2089319           180  02: Northeast        2.000000       2.000000   
2089320             0   01: Mid-West        6.000000       6.000000   
2089321             0        Unknown        6.000000       6.000000   
2089322            30      03: South        3.000000       3.000000   

        Claim_Count Count_NH Count_ALF Count_HHC Count_Unk  
0                 0        0         0         0         0  
1                 0        0         0         0         0  
2                 0        0         0         0         0  
3                 0        0         0         0         0  
4                 0        0         0         0         0  
...             ...      ...       ...       ...       ...  
2089318           0        0         0         0         0  
2089319           0        0         0         0         0  
2089320           0        0         0         

In [3]:
termination_df = dataframes['termination']
termination_df

Gender Incurred_Age_Bucket Incurred_Year_Bucket Claim_Type  \
0       Female             Unknown            2009-2010        HCC   
1       Female             Unknown            2003-2004         NH   
2       Female             Unknown            2005-2006      Other   
3       Female             Unknown            2003-2004        ALF   
4       Female             Unknown               < 2001         NH   
...        ...                 ...                  ...        ...   
627185    Male               85-89            2007-2008         NH   
627186    Male               75-79            2007-2008        HCC   
627187  Female               80-84            2013-2014        HCC   
627188    Male               70-74            2001-2002        HCC   
627189  Female               85-89            2003-2004        ALF   

               Region Diagnosis_Category Claim_Duration  Exposure  Deaths  \
0             Unknown            Unknown             80        77       0   
1             Unknown            Unknown             86        49       1   
2             Unknown            Unknown             88        16       1   
3             Unknown            Unknown            114        29       1   
4             Unknown            Unknown              4        56       5   
...               ...                ...            ...       ...     ...   
627185        Unknown         05: Injury             33         2       0   
627186      03: South            Unknown             31        62       3   
627187   01: Mid-West         07: Stroke             23         6       0   
627188  02: Northeast            Unknown             30         7       0   
627189        Unknown            Unknown             32        95       1   

        Recovery  Terminations  Benefit_Expiry  Others_Terminations  
0              2             2               1                    0  
1              0             1               0                    0  
2              0             1               0                    0  
3              0             1               0                    0  
4              1             6               0                    0  
...          ...           ...             ...                  ...  
627185         0             0               0                    0  
627186         0             3               0                    0  
627187         0             0               0                    0  
627188         0             0               0                    0  
627189         0             1               0                    0  

[627190 rows x 13 columns]

## Data Cleaning Process for "Incidence Datasets"

### Inspect for the presence of outliers and missing values in the variable

#### 1. Check out the dataframe at first and make sure the unique vlaue of each variable is reasonable 

In [4]:
unique_values = {col: incidence_df[col].unique() for col in incidence_df.columns}
unique_values

{'Group_Indicator': array(['Group', 'Individual'], dtype=object),
 'Gender': array(['Female', 'Male'], dtype=object),
 'Issue_Age_Bucket': array(['55-59', '60-64', '50-54', '65-69', '70-74', ' 0-49', '75-79',
        '80-84', '85-89', 'Unknown', '90+'], dtype=object),
 'Incurred_Age_Bucket': array(['55-59', '70-74', '60-64', '65-69', '50-54', '75-79', ' 0-49',
        '80-84', '85-89', '90+', 'Unknown'], dtype=object),
 'Issue_Year_Bucket': array(['2003-2005', '1997-1999', '2000-2002', '1994-1996', '2006-2008',
        '1991-1993', ' < 1991', '2012-2014', '2009-2011', '2015-2016',
        'Unknown'], dtype=object),
 'Policy_Year': array([' 1-3 years', '10-12 years', '13-15 years', ' 4-6 years',
        ' 7-9 years', '15+ years'], dtype=object),
 'Marital_Status': array(['Single', 'Unknown', 'Married'], dtype=object),
 'Premium_Class': array(['Standard', 'Preferred', 'Substandard'], dtype=object),
 'Underwriting_Type': array(['Other', 'Unknown', 'Full underwriting'], dtype=object),
 'Co

#### 2. Checking for "Unknown" Values That Do Not Contribute Information
The dataset contains 627,190 rows and 13 columns. For actuarial analysis, it is essential that the variables 'Issue_Age_Bucket', 'Incurred_Age_Bucket', and 'Issue_Year_Bucket' contain valid information since they are crucial in evaluating claim incidences. Rows with 'Unknown' values in these variables do not contribute to the analysis and can potentially skew the results. Currently, there are 519 'Unknown' values in 'Issue_Age_Bucket', 481 in 'Incurred_Age_Bucket', and 431 in 'Issue_Year_Bucket'. Removing rows with 'Unknown' values in these three variables is a necessary step to ensure data quality and reliability of the analysis.


In [13]:
# Calculate the number of 'Unknown' values in the 'Issue_Age_Bucket' column
unknown_issue_age = (incidence_df['Issue_Age_Bucket'] == 'Unknown').sum()
print(f"Number of 'Unknown' values in 'Issue_Age_Bucket': {unknown_issue_age}")

Number of 'Unknown' values in 'Issue_Age_Bucket': 519


In [6]:
unknown_incurred_age = (incidence_df['Incurred_Age_Bucket'] == 'Unknown').sum()
print(f"Number of 'Unknown' values in 'Incurred_Age_Bucket': {unknown_incurred_age}")


Number of 'Unknown' values in 'Incurred_Age_Bucket': 481


In [7]:
unknown_issue_year = (incidence_df['Issue_Year_Bucket'] == 'Unknown').sum()
print(f"Number of 'Unknown' values in 'Issue_Year_Bucket': {unknown_issue_year}")

Number of 'Unknown' values in 'Issue_Year_Bucket': 431


#### 3. Eliminate rows where below three variables have 'Unknown' values
- Counts of 'Unknown' values in key variables:
- 'Issue_Age_Bucket': 519
- 'Incurred_Age_Bucket': 481
- 'Issue_Year_Bucket': 431


In [17]:
# Eliminate rows where these three variables have 'Unknown' values
filtered_incidence_df = incidence_df[
    (incidence_df['Issue_Age_Bucket'] != 'Unknown') & 
    (incidence_df['Incurred_Age_Bucket'] != 'Unknown') & 
    (incidence_df['Issue_Year_Bucket'] != 'Unknown')
]

# Display the shape of the dataframe after elimination
print(f"Dataframe shape after elimination: {filtered_incidence_df.shape}")

Dataframe shape after elimination: (2088757, 31)


In [18]:
filtered_incidence_df

Group_Indicator  Gender Issue_Age_Bucket Incurred_Age_Bucket  \
0                 Group  Female            55-59               55-59   
1            Individual  Female            60-64               70-74   
2                 Group  Female            50-54               60-64   
3            Individual  Female            55-59               70-74   
4            Individual    Male            60-64               65-69   
...                 ...     ...              ...                 ...   
2089318      Individual    Male            55-59               60-64   
2089319      Individual    Male            65-69               70-74   
2089320      Individual  Female            65-69               75-79   
2089321      Individual  Female            50-54               55-59   
2089322      Individual  Female             0-49               50-54   

        Issue_Year_Bucket  Policy_Year Marital_Status Premium_Class  \
0               2003-2005    1-3 years         Single      Standard   
1               1997-1999  10-12 years         Single     Preferred   
2               2000-2002  10-12 years        Unknown      Standard   
3               1994-1996  13-15 years        Married      Standard   
4               2003-2005    4-6 years        Married   Substandard   
...                   ...          ...            ...           ...   
2089318         2012-2014    1-3 years        Married   Substandard   
2089319         1997-1999    7-9 years        Married      Standard   
2089320         1994-1996    7-9 years         Single     Preferred   
2089321         1994-1996    7-9 years         Single      Standard   
2089322         1997-1999    1-3 years         Single     Preferred   

         Underwriting_Type Coverage_Type_Bucket  ... ALF_EP_Bucket  \
0                    Other        Comprehensive  ...             0   
1                    Other        Comprehensive  ...             0   
2                  Unknown        Comprehensive  ...             0   
3                    Other        Comprehensive  ...             0   
4        Full underwriting        Comprehensive  ...            90   
...                    ...                  ...  ...           ...   
2089318  Full underwriting        Comprehensive  ...             0   
2089319  Full underwriting        Comprehensive  ...           180   
2089320              Other        Comprehensive  ...             0   
2089321            Unknown        Comprehensive  ...             0   
2089322  Full underwriting        Comprehensive  ...            90   

        HHC_EP_Bucket         Region Active_Exposure Total_Exposure  \
0                   0        Unknown       37.583332      37.583332   
1                   0  02: Northeast       14.500000      14.500000   
2                   0        Unknown     2250.000000    2252.000000   
3                   0   01: Mid-West       19.000000      19.000000   
4                  20  02: Northeast        2.000000       2.000000   
...               ...            ...             ...            ...   
2089318             0   01: Mid-West        7.083333       7.083333   
2089319           180  02: Northeast        2.000000       2.000000   
2089320             0   01: Mid-West        6.000000       6.000000   
2089321             0        Unknown        6.000000       6.000000   
2089322            30      03: South        3.000000       3.000000   

        Claim_Count Count_NH Count_ALF Count_HHC Count_Unk  
0                 0        0         0         0         0  
1                 0        0         0         0         0  
2                 0        0         0         0         0  
3                 0        0         0         0         0  
4                 0        0         0         0         0  
...             ...      ...       ...       ...       ...  
2089318           0        0         0         0         0  
2089319           0        0         0         0         0  
2089320           0        0         0         

#### 4. Inspect Underwriting_Type with Unknown value

In [12]:
unknown_Underwriting_Type = (incidence_df['Underwriting_Type'] == 'Unknown').sum()
print(f"Number of 'Unknown' values in 'Region': {unknown_Underwriting_Type}")

Number of 'Unknown' values in 'Region': 465920


#### 5. Inspect Region with Unknown value

In [16]:
unknown_Region = (incidence_df['Region'] == 'Unknown').sum()
print(f"Number of 'Unknown' values in 'Region': {unknown_Region}")

Number of 'Unknown' values in 'Region': 388271


### Impute Region with Conditional Mode Imputation

#### How Conditional Mode Imputation Helps Incidence Datasets:

1. **Enhances Data Integrity and Completeness**: In incidence datasets, 'Unknown' values in the 'Region' column can lead to information gaps or biases during analysis. Conditional Mode Imputation, by leveraging relationships known from other variables in the dataset (such as 'Policy_Year', 'Issue_Age_Bucket', etc.), allows for reasoned inference and filling of these unknown regional values. This not only adds to the dataset's completeness but also maintains the intrinsic correlation and consistency among the data.

2. **Reflects Actual Regional Distribution Characteristics**: By grouping similar records and calculating the modal value of 'Region' for each group based on conditions, Conditional Mode Imputation can more accurately reflect the regional distribution under specific conditions or backgrounds. This approach provides a closer approximation to reality, especially when the distribution of regions strongly correlates with specific variables, offering a more detailed view than simply replacing 'Unknown' with the overall most common region.

#### Improvement in Analysis Performance:

1. **Increases Model Accuracy and Interpretability**: A dataset with 'Unknown' values filled provides a more complete picture for analysis models, helping them capture correlations and patterns more accurately. This not only boosts the accuracy of analyses but also enhances the interpretability of model predictions or conclusions, as specific characteristics and trends of each region can be better identified and understood.

2. **Enhances Reliability of Decision Support**: For tasks that require risk assessment, product design, or market analysis based on incidence datasets, a more complete dataset reflecting the actual distribution of regions offers stronger support. By reducing the uncertainty associated with data missingness, decisions made on such data become more reliable.

3. **Avoids Analysis Bias**: Excluding or overlooking records with 'Unknown' regions could lead to overrepresentation or neglect of certain regions or demographics, introducing analysis bias. Conditional Mode Imputation allows these records to be included in the analysis but in a way that maintains data consistency and relevance, thereby avoiding potential biases.

#### Why Policy Year Matters:

The 'Policy_Year' is a significant variable in incidence datasets for several reasons:

1. **Temporal Trends**: Policy year can reflect changes in policies, market conditions, or risk profiles over time. Understanding these temporal trends is crucial for accurate risk assessment and forecasting.

2. **Regional Variations**: Policy uptake and preferences can vary by region in different years, influenced by economic conditions, regulatory changes, or cultural factors. Analyzing data based on policy year allows for a nuanced understanding of these regional variations.

3. **Product Evolution**: Insurance products and their features might evolve over the years. The policy year helps in distinguishing between different product generations and their impact on incidences.

4. **Risk Management**: Risk profiles can change over time due to various factors, including changes in underwriting practices or shifts in demographic patterns. The policy year provides a temporal dimension to analyze these changes, enabling more tailored risk management strategies.

In summary, Conditional Mode Imputation provides a method that ensures data integrity while preserving intrinsic data relationships. The inclusion of variables like 'Policy_Year' enriches the analysis by incorporating temporal and contextual dimensions, making the imputation process and subsequent analyses more aligned with real-world conditions and trends.

In [22]:
import numpy as np

# Step 1: Replace 'Unknown' with np.nan for easier handling with pandas methods
filtered_incidence_df['Region'].replace('Unknown', np.nan, inplace=True)

# Step 2 & 3: Impute 'Unknown' (now np.nan) in 'Region' based on the mode of each 'Policy_Year' group
for policy_year in filtered_incidence_df['Policy_Year'].unique():
    # Compute the mode of 'Region' for the current 'Policy_Year'
    mode_region = filtered_incidence_df.loc[filtered_incidence_df['Policy_Year'] == policy_year, 'Region'].mode()
    if not mode_region.empty:
        # Impute 'Unknown' values with the mode for this subset
        filtered_incidence_df.loc[(filtered_incidence_df['Policy_Year'] == policy_year) & (filtered_incidence_df['Region'].isna()), 'Region'] = mode_region[0]

# If there are still any 'Unknown' (np.nan) values left, fill them with the overall mode
overall_mode = filtered_incidence_df['Region'].mode()[0]
filtered_incidence_df['Region'].fillna(overall_mode, inplace=True)

# Check the imputation result
print(filtered_incidence_df['Region'].value_counts())


/var/folders/fs/5ct03tj50v1_hbz7dk31c6mw0000gn/T/ipykernel_2098/771386354.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  filtered_incidence_df['Region'].replace('Unknown', np.nan, inplace=True)


Region
03: South        933136
01: Mid-West     476202
04: West         416651
02: Northeast    257988
05: Other          4780
Name: count, dtype: int64


/var/folders/fs/5ct03tj50v1_hbz7dk31c6mw0000gn/T/ipykernel_2098/771386354.py:16: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  filtered_incidence_df['Region'].fillna(overall_mode, inplace=True)


In [23]:
filtered_incidence_df

Group_Indicator  Gender Issue_Age_Bucket Incurred_Age_Bucket  \
0                 Group  Female            55-59               55-59   
1            Individual  Female            60-64               70-74   
2                 Group  Female            50-54               60-64   
3            Individual  Female            55-59               70-74   
4            Individual    Male            60-64               65-69   
...                 ...     ...              ...                 ...   
2089318      Individual    Male            55-59               60-64   
2089319      Individual    Male            65-69               70-74   
2089320      Individual  Female            65-69               75-79   
2089321      Individual  Female            50-54               55-59   
2089322      Individual  Female             0-49               50-54   

        Issue_Year_Bucket  Policy_Year Marital_Status Premium_Class  \
0               2003-2005    1-3 years         Single      Standard   
1               1997-1999  10-12 years         Single     Preferred   
2               2000-2002  10-12 years        Unknown      Standard   
3               1994-1996  13-15 years        Married      Standard   
4               2003-2005    4-6 years        Married   Substandard   
...                   ...          ...            ...           ...   
2089318         2012-2014    1-3 years        Married   Substandard   
2089319         1997-1999    7-9 years        Married      Standard   
2089320         1994-1996    7-9 years         Single     Preferred   
2089321         1994-1996    7-9 years         Single      Standard   
2089322         1997-1999    1-3 years         Single     Preferred   

         Underwriting_Type Coverage_Type_Bucket  ... ALF_EP_Bucket  \
0                    Other        Comprehensive  ...             0   
1                    Other        Comprehensive  ...             0   
2                  Unknown        Comprehensive  ...             0   
3                    Other        Comprehensive  ...             0   
4        Full underwriting        Comprehensive  ...            90   
...                    ...                  ...  ...           ...   
2089318  Full underwriting        Comprehensive  ...             0   
2089319  Full underwriting        Comprehensive  ...           180   
2089320              Other        Comprehensive  ...             0   
2089321            Unknown        Comprehensive  ...             0   
2089322  Full underwriting        Comprehensive  ...            90   

        HHC_EP_Bucket         Region Active_Exposure Total_Exposure  \
0                   0      03: South       37.583332      37.583332   
1                   0  02: Northeast       14.500000      14.500000   
2                   0      03: South     2250.000000    2252.000000   
3                   0   01: Mid-West       19.000000      19.000000   
4                  20  02: Northeast        2.000000       2.000000   
...               ...            ...             ...            ...   
2089318             0   01: Mid-West        7.083333       7.083333   
2089319           180  02: Northeast        2.000000       2.000000   
2089320             0   01: Mid-West        6.000000       6.000000   
2089321             0      03: South        6.000000       6.000000   
2089322            30      03: South        3.000000       3.000000   

        Claim_Count Count_NH Count_ALF Count_HHC Count_Unk  
0                 0        0         0         0         0  
1                 0        0         0         0         0  
2                 0        0         0         0         0  
3                 0        0         0         0         0  
4                 0        0         0         0         0  
...             ...      ...       ...       ...       ...  
2089318           0        0         0         0         0  
2089319           0        0         0         0         0  
2089320           0        0         0         

## Data Cleaning Process for "Termination Datasets"

## Split both "Incidence" and "Termination" datasets into train/test

In [82]:
import pandas as pd
from sklearn.model_selection import train_test_split

# Now we will split the 'incidence' DataFrame into training and validation sets
train__filtered_incidence_df, validation_filtered_incidence_df = train_test_split(filtered_incidence_df, test_size=0.2, random_state=42)

# train_df is now your training set, and validation_df is your validation set.

In [85]:
train__filtered_incidence_df

Group_Indicator  Gender Issue_Age_Bucket Incurred_Age_Bucket  \
1702737      Individual  Female            50-54               60-64   
1380059      Individual  Female            70-74               80-84   
837211       Individual  Female            60-64               60-64   
1315575      Individual  Female            80-84               80-84   
907204       Individual  Female            55-59               55-59   
...                 ...     ...              ...                 ...   
259245            Group  Female            65-69               65-69   
1414796      Individual  Female            55-59               65-69   
131965       Individual    Male            70-74               80-84   
671338       Individual  Female            70-74               85-89   
121987       Individual    Male            70-74               85-89   

        Issue_Year_Bucket  Policy_Year Marital_Status Premium_Class  \
1702737         2003-2005    7-9 years         Single     Preferred   
1380059         2000-2002  13-15 years        Married      Standard   
837211          2000-2002    4-6 years        Unknown      Standard   
1315575         1994-1996    4-6 years        Married      Standard   
907204          1994-1996    4-6 years        Married      Standard   
...                   ...          ...            ...           ...   
259245          2003-2005    4-6 years        Married      Standard   
1414796         1997-1999  10-12 years         Single      Standard   
131965          2009-2011    7-9 years        Unknown      Standard   
671338          1997-1999    15+ years         Single     Preferred   
121987          2000-2002  13-15 years        Married     Preferred   

         Underwriting_Type Coverage_Type_Bucket  ... ALF_EP_Bucket  \
1702737              Other        Comprehensive  ...            90   
1380059            Unknown        Comprehensive  ...             0   
837211             Unknown        Comprehensive  ...            90   
1315575  Full underwriting                Other  ...             0   
907204   Full underwriting                Other  ...            30   
...                    ...                  ...  ...           ...   
259245               Other        Comprehensive  ...             0   
1414796            Unknown        Comprehensive  ...             0   
131965   Full underwriting        Comprehensive  ...             0   
671338             Unknown        Comprehensive  ...             0   
121987             Unknown                Other  ...             0   

        HHC_EP_Bucket         Region Active_Exposure Total_Exposure  \
1702737             0       04: West       19.583332      19.583332   
1380059             0      03: South       84.333336     108.916660   
837211              0      03: South        3.000000       3.000000   
1315575             0      03: South        3.750000       3.750000   
907204              0      03: South        2.583333       2.583333   
...               ...            ...             ...            ...   
259245              0  02: Northeast        4.000000       4.000000   
1414796             0       04: West       62.750000      62.833332   
131965              0  02: Northeast        2.416666       2.416666   
671338              0      03: South       14.249999      19.916664   
121987              0       04: West        3.083333       3.083333   

        Claim_Count Count_NH Count_ALF Count_HHC Count_Unk  
1702737           0        0         0         0         0  
1380059           8        0         0         8         0  
837211            0        0         0         0         0  
1315575           0        0         0         0         0  
907204            0        0         0         0         0  
...             ...      ...       ...       ...       ...  
259245            0        0         0         0         0  
1414796           1        0         0         1         0  
131965            0        0         0         

In [84]:
validation_filtered_incidence_df

Group_Indicator  Gender Issue_Age_Bucket Incurred_Age_Bucket  \
1756216           Group  Female            50-54               55-59   
2001415      Individual  Female            60-64               75-79   
454164            Group    Male             0-49                0-49   
1073250      Individual    Male            65-69               80-84   
811617       Individual    Male            65-69               65-69   
...                 ...     ...              ...                 ...   
736427       Individual  Female            60-64               75-79   
506473       Individual  Female            65-69               65-69   
1894427      Individual  Female             0-49               60-64   
244198            Group  Female            55-59               60-64   
2069551      Individual  Female            60-64               65-69   

        Issue_Year_Bucket  Policy_Year Marital_Status Premium_Class  \
1756216         2009-2011    7-9 years        Unknown      Standard   
2001415         1994-1996    15+ years        Married      Standard   
454164          1997-1999  10-12 years        Married      Standard   
1073250         1997-1999  13-15 years        Married      Standard   
811617          1994-1996    4-6 years        Married      Standard   
...                   ...          ...            ...           ...   
736427          2000-2002    15+ years        Married     Preferred   
506473          2000-2002    4-6 years        Unknown   Substandard   
1894427         2000-2002    15+ years         Single     Preferred   
244198          2006-2008    1-3 years        Unknown      Standard   
2069551         1997-1999    7-9 years        Married     Preferred   

         Underwriting_Type Coverage_Type_Bucket  ... ALF_EP_Bucket  \
1756216            Unknown        Comprehensive  ...             0   
2001415  Full underwriting        Comprehensive  ...             0   
454164               Other        Comprehensive  ...             0   
1073250              Other        Comprehensive  ...             0   
811617             Unknown        Comprehensive  ...             0   
...                    ...                  ...  ...           ...   
736427               Other        Comprehensive  ...            90   
506473   Full underwriting                Other  ...             0   
1894427            Unknown        Comprehensive  ...             0   
244198   Full underwriting        Comprehensive  ...             0   
2069551              Other        Comprehensive  ...            20   

        HHC_EP_Bucket         Region Active_Exposure Total_Exposure  \
1756216             0      03: South       32.749996      32.749996   
2001415             0  02: Northeast        2.000000       2.000000   
454164              0  02: Northeast        7.000000       7.000000   
1073250             0   01: Mid-West       63.583332      65.583328   
811617              0  02: Northeast       23.416662      23.416662   
...               ...            ...             ...            ...   
736427             30   01: Mid-West        1.500000       1.500000   
506473              0   01: Mid-West        3.000000       3.000000   
1894427             0      03: South        2.083333       2.083333   
244198              0      03: South        2.000000       2.000000   
2069551            20       04: West        9.000000       9.000000   

        Claim_Count Count_NH Count_ALF Count_HHC Count_Unk  
1756216           0        0         0         0         0  
2001415           0        0         0         0         0  
454164            0        0         0         0         0  
1073250           3        1         1         1         0  
811617            0        0         0         0         0  
...             ...      ...       ...       ...       ...  
736427            0        0         0         0         0  
506473            0        0         0         0         0  
1894427           0        0         0         

In [5]:
# Now we will split the 'incidence' DataFrame into training and validation sets

train_termination_df, validation_termination_df = train_test_split(termination_df, test_size=0.2, random_state=42)